firedrake/petsc.py

```diff
@@ -249,3 +256,70 @@ def inserted_options(self):
         finally:
             for k in self.to_delete:
                 del self.options_object[self.options_prefix + k]
+
+
+def _extract_comm(obj):
+    comm = None
+    if isinstance(obj, (PETSc.Comm, mpi.MPI.Comm)):
+        try:
+            if mpi.is_pyop2_comm(obj):
+                comm = obj
+            else:
+                internal_comm = obj.Get_attr(mpi.innercomm_keyval)
```

Comment:
Is there a nice PyOP2 function for this?

Reply:
Yes, but it increases the reference count. In this instance, since we do not return the communicator, I do not think that is necessary. However, when I fix [PyOP2 issue 702](https://github.com/OP2/PyOP2/issues/702) I can duplicate it and let the weakref finalizer clean things up.

Reply:
I'm unresolving this until I fix https://github.com/OP2/PyOP2/issues/702